uqcsbot/starboard.py

```diff
@@ -1,244 +1,598 @@
+                elif entry.sent is None:
+                    raise BlacklistedMessageError()
+
+                channel = self.bot.get_channel(channel_id)
+
+                return (
+                    await self._fetch_message_or_none(channel, message_id),
+                    await self._fetch_message_or_none(
+                        self.starboard_channel, entry.sent
+                    ),
+                )
+            else:
+                channel = self.bot.get_channel(channel_id)
+
+                return (
+                    await self._fetch_message_or_none(channel, message_id),
+                    None,
+                )
+
+    async def _count_num_reacts(
+        self, data: Tuple[discord.Message | None, discord.Message | None]
+    ) -> int:
+        """Takes _lookup_from_id data (two messages, maybe None).
+
+        Returns the number of unique reactions across both messages, not including the authors of those messages.
+        """
+        users = []
+        authors = []
+
+        for message in data:
+            if message is None:
+                continue
+
+            # store the message authors so we can discard their reacts later
+            # grandfathering old messages where their reacts were not auto-deleted, also failsafes are nice, etc
+            authors += [message.author.id]
+            reaction = discord.utils.get(message.reactions, emoji=self.starboard_emoji)
+            if reaction is not None:
+                # we use the user.id to describe the reaction so we can set() it and
+                # eliminate duplicates (only count one reaction per person)
+                users += [user.id async for user in reaction.users()]
+
+        return len(set([user for user in users if user not in authors]))
+
+    def _generate_message_text(
+        self, reaction_count: int, recieved_msg: (discord.Message | None)
+    ) -> str:
+        return (
+            f"{str(self.starboard_emoji)} {reaction_count} | "
+            f"{('#' + recieved_msg.channel.name) if recieved_msg is not None else 'OoOoO... ghost message!'}"
+        )
+
+    def _generate_message_embeds(
+        self, recieved_msg: discord.Message
+    ) -> List[discord.Embed]:
+        """Creates the starboard embed for a message, including author, colours, replies, etc."""
+        embed = discord.Embed(
+            color=recieved_msg.author.top_role.color, description=recieved_msg.content
+        )
+        embed.set_author(
+            name=recieved_msg.author.display_name,
+            icon_url=recieved_msg.author.display_avatar.url,
+        )
+        embed.set_footer(
+            text=recieved_msg.created_at.astimezone(tz=self.BRISBANE_TZ).strftime(
+                "%b %d, %H:%M:%S"
+            )
+        )
+
+        if len(recieved_msg.attachments) > 0:
+            embed.set_image(url=recieved_msg.attachments[0].url)
             # only takes the first attachment to avoid sending large numbers of images to starboard.
```

Comment:
Is this still required with the new discord gallery view?

Reply:
I still think it's worth reducing image-spam on the starboard anyway. How often does a message with >2 attachments get sent to the discord, and how often would that message then get starred?

Reply:
Good point. Up to you